v2/tools/generator/internal/codegen/pipeline/detect_skipping_properties.go

```diff
@@ -0,0 +1,242 @@
+// these properties and can raise an error if a specific property appears more than once (implying there are two or more
+// sequence versions of that property).
+//
+// To illustrate, assume we have the following set of objects across versions of our mythical CRM service:
+//
+// v1: Person(GivenName, FamilyName, Address)
+// v2: Person(KnownAs, FullName, FamilyName)
+// v3: Person(KnownAs, FullName, FamilyName, Address)
+//
+// Scanning these types, we find:
+//
+// FamilyName: present in (v1, v2, v3); no issue
+// FullName: present in (v2, v3); no issue
+// KnownAs: also present in (v2, v3); no issue
+// GivenName: present in (v1); no issue
+//
+// Address: present in (v1, v2); issue to report, aborting the pipeline.
+//
+// Additional complexities:
+// - We need to handle type renaming between versions.
+// - When introduced, we will also need to handle property renaming between versions
+//
+func DetectSkippingProperties() Stage {
+	return MakeStage(
+		DetectSkippingPropertiesStageID,
+		"Detect properties that skip resource or object versions",
+		func(ctx context.Context, state *State) (*State, error) {
+			detector := NewSkippingPropertyDetector(state.Types(), state.ConversionGraph())
+
+			// Add resources and objects to the graph
+			for _, def := range state.Types() {
+				if t, ok := astmodel.AsPropertyContainer(def.Type()); ok {
+					err := detector.AddProperties(def.Name(), t)
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+
+			err := detector.CheckForSkippedProperties()
+
+			return state, err
+		})
+}
+
+type SkippingPropertyDetector struct {
+	links              map[astmodel.PropertyReference]astmodel.PropertyReference // Individual links in chains of related properties
+	observedProperties *astmodel.PropertyReferenceSet                            // Set of properties we've observedProperties
+	types              astmodel.Types                                            // Set of all known types
+	conversionGraph    *storage.ConversionGraph                                  // Graph of conversions between types
+}
+
+// NewSkippingPropertyDetector creates a new graph for tracking chains of properties as they evolve through different
+// versions of a resource or object.
+// types is a set of all known types.
+// conversionGraph contains all of the conversions/transitions between versions.
+func NewSkippingPropertyDetector(types astmodel.Types, conversionGraph *storage.ConversionGraph) *SkippingPropertyDetector {
+	return &SkippingPropertyDetector{
+		links:              make(map[astmodel.PropertyReference]astmodel.PropertyReference),
+		observedProperties: astmodel.NewPropertyReferenceSet(),
+		types:              types,
+		conversionGraph:    conversionGraph,
+	}
+}
+
+// AddProperties adds all the properties from the specified type to the graph.
+func (detector *SkippingPropertyDetector) AddProperties(
+	name astmodel.TypeName,
+	container astmodel.PropertyContainer) error {
```

Comment:
This could take a `[]*astmodel.PropertyDefinition` and you could move the `container.Properties()` call into the caller, so that the call was: `err := detector.AddProperties(def.Name(), t.Properties())`

That might make unit testing a bit easier as you wouldn't need an actual `PropertyContainer` instance passed in here, and it makes `AddProperties` and `AddProperty` more symmetric (as both then take the same types, just one is a collection and one is a singlet).

Overall obviously quite minor though.

Reply:
Good suggestion.